lib/preview/previewUtils.js

```diff
@@ -0,0 +1,183 @@
+  }
+  const endOfBodyIndex = html.lastIndexOf("</body>");
+  return insertAt(html, endOfBodyIndex, script);
+}
+
+const addRefreshScript = (html) => {
+  const refreshScript = `
+  <script>
+  (() => {
+    const NORMAL_WAIT_MS = 1000;
+    const BACKOFF_RATIO = 2;
+    let nextWait = NORMAL_WAIT_MS;
+    let timeWaited = 0;
+
+    setInterval(() => {
+      timeWaited += NORMAL_WAIT_MS;
+      if (timeWaited !== nextWait) return;
+      timeWaited = 0;
+      const res = fetch(window.location.origin + '/meta')
+        .then(async (res) => {
+          const hsServerState = await res.json();
+          if (hsServerState["REMOTE_FS_IS_DIRTY"]) {
+            location.reload();
+          }
+          nextWait = NORMAL_WAIT_MS;
+        })
+        .catch(err => {
+          nextWait *= BACKOFF_RATIO;
+          console.log('Disconnected from local server... (retrying in ' + nextWait / 1000 + 's)');
+        });
+    }, NORMAL_WAIT_MS);
+  })();
+  </script>
+  `;
+  return insertAtEndOfBody(html, refreshScript);
+}
+
+const getSubDomainFromValidLocalDomain = hostname => {
+  for (const validProxyDomainSuffix of VALID_PROXY_DOMAIN_SUFFIXES) {
+    if (hostname.endsWith(`.${validProxyDomainSuffix}`)) {
+      return hostname.slice(0, -1 * validProxyDomainSuffix.length - 1);
+    }
+  }
+  return undefined;
+};
+
+// From(ish) https://git.hubteam.com/HubSpot/cloudflare-workers/blob/master/worker-lib/src/Constants.ts
+const internalRoutes = {
+  HCMS: '/_hcms/',
+  HS_FS: '/hs-fs/',
+  HUB_FS: '/hubfs/'
+}
+
+const isInternalCMSRoute = (req) =>
+  Object.values(internalRoutes).some((route => req.path.startsWith(route)));
+
+const silenceConsoleWhile = async (act, ...args) => {
+  const tmpConsole = console;
+  console = { log: () => {} }
+  const result = await act(...args);
+  console = tmpConsole;
+  return result;
+}
+
+const memoize = (func, cacheBustCallback) => {
+  const cache = {};
+  return async (...args) => {
+    const stringArgs = args.toString();
+    const storedResult = cache[stringArgs];
+    if (storedResult && (cacheBustCallback ? !cacheBustCallback() : true)) {
+      //console.log(`Cache hit ${func}`)
+      return storedResult;
+    }
+    //console.log(`Cache miss ${func}`)
+    const res = await func(...args);
+    cache[stringArgs] = res;
+    return res;
+  }
+}
+
+const hidePreviewInDest = (previewDest) => previewDest.split('/').slice(2).join('/');
+
+const trackPreviewEvent = async (action) => {
+  if (!isTrackingAllowed()) {
+    return;
+  }
+  const accountId = getAccountId();
+
+  trackUsage(
+    'cli-interaction',
+    'INTERACTION',
+    {
+      applicationName: 'hubspot.preview',
+      os: `${platform()} ${release()}`,
+      authType: getAuthType(accountId),
+      action,
+    },
+    accountId
+  ).catch(
+    (err) => {
+      console.error(`trackUsage failed: ${JSON.stringify(err, null, 2)}`);
+    }
+  );
+}
+
+const getAuthType = (accountId) => {
```

Comment:
I'm curious why this defaults to `apikey `. That's a deprecated auth Type. I'd guess that `personalaccesskey` would make a better default (unless I'm missing something).

I also wonder if this util would be better off in the config file so it could be used in other places too.

Reply:
Ripped this straight from VSCE without thinking about it where we use it to include that in telemetry there too: https://github.com/HubSpot/hubspot-cms-vscode/blob/2629b847cc2dbd99b146bfe08a53f3f6e60ef9d9/src/lib/tracking.ts#L73

+1 on changing (imo 'unknown' as default) and moving it to config to share usage.